zebrad/src/commands/start/sync.rs

```diff
@@ -302,23 +296,49 @@ where
                         // It indicates that the remote peer does not have any blocks
                         // following the prospective tip.
                         match (hashes.first(), hashes.len()) {
-                            (_, 0) => {
-                                tracing::debug!("skipping empty response");
+                            (None, _) => {
+                                tracing::debug!("ExtendTips: skipping empty response");
                                 continue;
                             }
                             (_, 1) => {
-                                tracing::debug!("skipping length-1 response, in case it's an unsolicited inv message");
+                                tracing::debug!("ExtendTips: skipping length-1 response, in case it's an unsolicited inv message");
                                 continue;
                             }
                             (Some(hash), _) if (hash == &self.genesis_hash) => {
-                                tracing::debug!("skipping response, peer could not extend the tip");
+                                tracing::debug!(
+                                    "ExtendTips: skipping response, peer could not extend the tip"
+                                );
                                 continue;
                             }
-                            _ => {}
+                            (Some(&hash), _) => {
+                                // Check for hashes we've already seen.
+                                // This happens a lot near the end of the chain.
+                                let depth = self.get_depth(hash).await?;
+                                if let Some(depth) = depth {
+                                    tracing::debug!(
+                                        ?depth,
+                                        ?hash,
+                                        "ExtendTips: skipping response, peer returned a duplicate hash: already in state"
+                                    );
+                                    continue;
```

Comment:
Will this race with block downloads that haven't been committed to the state yet?

Reply:
Yes, and that's ok, because it's only a heuristic that reduces the number of duplicate block downloads.

If the state wins the race, then we skip some duplicate blocks.
If ExtendTips wins the race, then we download some duplicate blocks, and the verifiers or state ignore them.

Reply:
(I added a comment to clarify.)

Reply:
See c590c18123bb35d832a8b26fad3a1ae0e0c5f451